docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md

```diff
@@ -0,0 +1,174 @@
+    * TODO: A reduction in the list of CRDs to install should result in only the new list CRDs of CRDs getting new versions?
+      Or maybe it needs to actually upgrade those versions too even though they weren't specified, to maintain a consistent
+      storage version?
+
+### Question 1: How to identify the subset of resources to install by default?
+
+As mentioned above, a key goal is a reduction in the default number of CRDs installed. That means we somehow need to 
+choose which CRDs we will install and which CRDs we will not.
+
+#### Answer 1a: We handpick a set of resources we feel are "critical" and install just those by default
+
+Pros:
+* Has at least a chance of working for some user-scenarios out of the box.
+* Ensures that resources which are extremely likely to be needed (such as Microsoft.Resources/resourceGroup) are included 
+  by default.
+
+Cons:
+* Arbitrary.
+* May result in higher kube-apiserver memory consumption than is strictly needed for users as it is likely to include some
+  resources which they may not use.
+
+#### Answer 1b: We install nothing by default
+
+Pros:
+* Users choose exactly what they need.
+
+Cons:
+* The operator is nonfunctional out of the box, which is a poor new-user experience.
+
+#### Answer 1c: Selection of CRDs to install is required and the operator pod will not launch without it
+
+Can also use [Helm required values](https://stackoverflow.com/questions/53866196/how-best-to-say-a-value-is-required-in-a-helm-chart)
+to ensure that the value is required at the Helm chart level, although making it required may run afoul of requirement #3.
+
+Pros:
+* Users choose exactly what they need.
+
+Cons:
+* The operator is nonfunctional out of the box, but at least it's clearly nonfunctional and has a good error why.
+
+### Question 2: How do users specify what CRDs to install?
+
+#### Answer 2a: using our existing --crd-pattern command line parameter on the operator pod
+
+Pros:
+* Exists today
+
+Cons:
+* Command-line isn't preserved between upgrades, which fails requirement #3 (seamless upgrade experience).
+
+#### Answer 2b: Using a ConfigMap
+
+A `ConfigMap` would persist between upgrades and so in theory could be used to remember what was installed previously
+without the user needing to re-specify it.
+
+One challenge is that in the Helm context if the `ConfigMap` is created by Helm, Helm will overwrite it in 
+subsequent upgrades. This could be dealt with by using a 
+[pre-install hook](https://stackoverflow.com/questions/55503893/helm-patch-default-service-account). We would create and
+patch the `ConfigMap` in the pre-install hook.
+
+ConfigMap structure:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: managed-resources
+  namespace: azureserviceoperator-system
+data:
+  "all": ""  # magic key meaning "*" which isn't supported in keys
+  microsoft.resources: "*"
+  microsoft.storage: "StorageAccount,StorageAccountsBlobService"
```

Comment:
We might need handling somewhere to check if parent CRD is installed or not. As for in this example, what if a new user, does something like `microsoft.storage: "StorageAccountsBlobService"` and assume it'll include StorageAccount or if its not required.

Reply:
I'll add an open questions section and put this there. My current thinking on this is that we avoid this complexity and strongly recommend users install entire providers such as `compute/*`.

There are other links to consider too:
- Owners (i.e. resourceGroup owns a lot of things even though it's not in their group)
- Cross resource references: i.e. compute references networking. This is pretty involved to work out automatically and there's also no guarantee that just because a resource _can_ link to a Subnet or a LoadBalancer that the customer actually has that use-case.